### PR TITLE
ZipperBams to produce mate score ("ms") for samtools markdup

### PR DESCRIPTION
**Pull request by nh13**
Fixes: #951

### CODE REVIEW

**Review by tfenne — CHANGES_REQUESTED** [review]
I don't object in principal, but these should all probably be of the form:

```scala
primary.get[Int]("AS").foreach(supp("ms") = _)
```

so that it will work for aligners that don't produce an `AS` tag and on pairs where only one read is aligned.

File: src/main/scala/com/fulcrumgenomics/bam/Bams.scala [hunk 1]
```diff
@@ -105,16 +105,21 @@ case class Template(r1: Option[SamRecord],
...
+      supp("ms")  = primary[Int]("AS")
     }
     for (primary <- r2; supp <- r1Supplementals) {
       SamPairUtil.setMateInformationOnSupplementalAlignment(supp.asSam, primary.asSam, true)
+      supp("ms")  = primary[Int]("AS")
     }
     for (first <- r1; second <- r2) {
       SamPairUtil.setMateInfo(first.asSam, second.asSam, true)
+      first("ms")  = second[Int]("AS")
+      second("ms") = first[Int]("AS")
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/bam/Bams.scala:122]:
```suggestion
  /** Fixes mate information and sets mate cigar and mate score on all primary and supplementary (but not secondary) records. */
  def fixMateInfo(): Unit = {
    // Developer note: the mate score ("ms") tag is used by samtools markdup
    for (primary <- r1; supp <- r2Supplementals) {
      SamPairUtil.setMateInformationOnSupplementalAlignment(supp.asSam, primary.asSam, true)
      primary.get[Int]("AS").foreach(supp("ms") = _))
    }
    for (primary <- r2; supp <- r1Supplementals) {
      SamPairUtil.setMateInformationOnSupplementalAlignment(supp.asSam, primary.asSam, true)
      primary.get[Int]("AS").foreach(supp("ms") = _))
    }
    for (first <- r1; second <- r2) {
      SamPairUtil.setMateInfo(first.asSam, second.asSam, true)
      first.get[Int]("AS").foreach(second("ms") = _))
      second.get[Int]("AS").foreach(first("ms") = _))
```

File: src/main/scala/com/fulcrumgenomics/bam/Bams.scala [hunk 2]
```diff
@@ -105,16 +105,21 @@ case class Template(r1: Option[SamRecord],
...
+      primary.get[Int]("AS").foreach(supp("ms") = _))
     }
     for (primary <- r2; supp <- r1Supplementals) {
       SamPairUtil.setMateInformationOnSupplementalAlignment(supp.asSam, primary.asSam, true)
+      primary.get[Int]("AS").foreach(supp("ms") = _))
     }
     for (first <- r1; second <- r2) {
       SamPairUtil.setMateInfo(first.asSam, second.asSam, true)
+      first.get[Int]("AS").foreach(second("ms") = _))
+      second.get[Int]("AS").foreach(first("ms") = _))
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/bam/Bams.scala:122]:
```suggestion
  /** Fixes mate information and sets mate cigar and mate score on all primary and supplementary (but not secondary) records. */
  def fixMateInfo(): Unit = {
    // Developer note: the mate score ("ms") tag is used by samtools markdup
    for (primary <- r1; supp <- r2Supplementals) {
      SamPairUtil.setMateInformationOnSupplementalAlignment(supp.asSam, primary.asSam, true)
      primary.get[Int]("AS").foreach(supp("ms") = _)
    }
    for (primary <- r2; supp <- r1Supplementals) {
      SamPairUtil.setMateInformationOnSupplementalAlignment(supp.asSam, primary.asSam, true)
      primary.get[Int]("AS").foreach(supp("ms") = _)
    }
    for (first <- r1; second <- r2) {
      SamPairUtil.setMateInfo(first.asSam, second.asSam, true)
      first.get[Int]("AS").foreach(second("ms") = _)
      second.get[Int]("AS").foreach(first("ms") = _)
```

**Review by tfenne — APPROVED** [review]
LGTM but could you add/amend a couple of tests to make sure it works on both reads that have and don't have AS set?

**Review by TedBrookings — APPROVED** [review]
Sorry for the delay. Looks good to me

**Pull request merged**